### DISCUSSION TITLE
`underive` does not delete the derived property unless there is a dependency defined

### DISCUSSION THREAD

**Discussion by dlindahl** [pmndrs/valtio#805] [Bug report]
## Summary
Calling `underive` with `delete: true` does not delete the derived property value unless there is a dependency.

1. Calling `derive` [subscribes to the dependencies](https://github.com/pmndrs/valtio/blob/ba53bf6fec891ec82a6fca211c562e49dbc38aae/src/vanilla/utils/derive.ts#L259) defined within the `derive` function.
2. Calling `underive` with `delete: true` [loops through all of the dependencies](https://github.com/pmndrs/valtio/blob/ba53bf6fec891ec82a6fca211c562e49dbc38aae/src/vanilla/utils/derive.ts#L298).
3. After removing each dependency subscription, [the key of the subscription is added to a `Set` to delete later](https://github.com/pmndrs/valtio/blob/ba53bf6fec891ec82a6fca211c562e49dbc38aae/src/vanilla/utils/derive.ts#L303).
4. Once each subscription has been removed, [the derived functions that matched a subscription key are deleted from the `proxyObject`](https://github.com/pmndrs/valtio/blob/ba53bf6fec891ec82a6fca211c562e49dbc38aae/src/vanilla/utils/derive.ts#L308-L310).

The problem arises from a situation where the derived function never calls `get` (for whatever reason) which prevents a subscription from being created. The lack of a subscription prevents the property from being deleted which then prevents new calls to `derive` because the property is already defined.

Is this silly? Yes.
Was this hard to debug. Also yes. 😄 

Property deletion should probably be decoupled from the existence of a subscription, especially since I see some async subscriptions being handled in the `derive` code which could possibly introduce a situation where `underive` is called before the async value resolves which would result in a similar error. 

## Link to reproduction

```js
derive({
  myProperty(get) {
    return 123; // NOTE: Not using `get` which does not create a dependency.
  }
}, {
  proxy: myProxy
})

underive(myProxy, { delete: true, keys: ['myProperty'])
// Throws Error: object property already defined
```

Pardon the lack of a codesandbox repro as the problem seemed pretty straight forward. I can make one if you really need one though.

## Check List

Please do not ask questions in issues.

- [ ] ~I've already opened a [discussion](https://github.com/pmndrs/valtio/discussions) before opening this issue, or already discussed in other media.~

Please include a minimal reproduction.

- [ ] ~I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.~

**Comment by dai-shi** [comment]
Nice catch. Thanks for reporting.

Would you or anyone like to fix it? `derive` is already complicated, and hope to simplify the implementation.

**Comment by dai-shi** [comment]
Here's a new proposition, which might be controversial: #792

**Comment by dai-shi** [comment]
Transferred to: https://github.com/valtiojs/derive-valtio/issues/1
